### PR TITLE
updated offer feature

### PR DESCRIPTION
**Pull request by emstryszawski**

### CODE REVIEW

File: src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java [hunk 1]
```diff
@@ -35,36 +41,79 @@ public OfferService(OfferRepository offerRepository, FeatureRepository featureRe
...
+        }
     }
 
     public OfferDto updateOffer(long id, CreateOfferDto offerDto) {
-        Offer offer = offerRepository.getById(id);
-        modelMapper.map(offerDto, offer);
-        Offer updatedOffer = offerRepository.save(offer);
-        return modelMapper.map(updatedOffer, OfferDto.class);
+        Optional<Offer> optionalOffer = offerRepository.findById(id);
+        if (authService.isPoLoggedIn() || authService.isAdminLoggedIn()) {
```
Comment (MaciejY0R5) [inline src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: lepiej by bylo to sprawdzac privilege, czyli ze kazda rola, ma przypisane liste privilege
i np. jest privilege CAN_UPDATE_OFFER 

dzieki temu, jesli cos sie zmieni i np. admin ma nie miec mozliwosci updatowac offerty, to wtedy zabieracie mu ten privilege z bazy i juz tego nie moze robic, a przy obecnym rozwiazaniu muscie zmienic ten kod ;)

**Review by MaciejY0R5 — CHANGES_REQUESTED** [review]

File: src/main/java/dk/bec/polonez/reservationsystem/controller/CustomExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@ControllerAdvice
+public class CustomExceptionHandler extends ResponseEntityExceptionHandler
+{
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/controller/CustomExceptionHandler.java]: **advice** Minor: Weird code formatting, shouldn't ```{``` be in line 19? (ALT + CTRL + L in IntelliJ)

File: src/main/java/dk/bec/polonez/reservationsystem/dto/offerDto/CreateOfferDto.java [hunk 3]
```diff
@@ -2,12 +2,15 @@
 
 import lombok.Data;
 
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.Size;
+
 @Data
 public class CreateOfferDto {
 
+    @NotBlank(message = "Name is mandatory")
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/dto/offerDto/CreateOfferDto.java:12]: **advice** What about size validation for name?

File: src/main/java/dk/bec/polonez/reservationsystem/exception/ErrorResponse.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class ErrorResponse {
+
+    private String status;
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/exception/ErrorResponse.java]: **question** Is it possible to use ```enum``` for status?

File: src/main/java/dk/bec/polonez/reservationsystem/model/Offer.java [hunk 5]
```diff
@@ -25,6 +25,10 @@
     @JoinColumn
     private User owner;
 
+    private double pricePerUnit;
+
+    private String unitDuration;
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/model/Offer.java]: **advice** I can see this field means some duration (okres trwania) - is it possible to use ```Duration``` type from ```java.time```?

File: src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java [hunk 6]
```diff
@@ -35,36 +41,79 @@ public OfferService(OfferRepository offerRepository, FeatureRepository featureRe
...
-        Offer offer = modelMapper.map(offerDto, Offer.class);
-        Offer savedOffer = offerRepository.save(offer);
-        return modelMapper.map(savedOffer, OfferDto.class);
+        if (authService.isPoLoggedIn() || authService.isAdminLoggedIn()) {
+            Offer offer = modelMapper.map(offerDto, Offer.class);
+            offer.setOwner(authService.getCurrentUser());
+            Offer savedOffer = offerRepository.save(offer);
+            return modelMapper.map(savedOffer, OfferDto.class);
+        } else {
+            throw new AuthenticationException(HttpStatus.FORBIDDEN, "Access denied, log as Place Owner or System Admin");
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: **advice** Business exception shoule be thrown here. ```OfferService``` should abstract away from REST and HTTP-specific details.

Comment (emstryszawski) [reply src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: Not sure If I understood. Where am I supposed to throw exception then?

Comment (mgjvm) [reply src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: My comment was similar to our comments in previous pull-requests, that's why I didn't explain fully, sorry. Generally you should throw your own custom exception in your service (like ```NoAccessToOfferCreationException```) and then you should implement custom exception mapper (in a separate class) that would map your custom exception to specific HTTP error code. By doing this we are separating pure business logic from HTTP/REST details.

File: src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java [hunk 7]
```diff
@@ -35,36 +44,86 @@ public OfferService(OfferRepository offerRepository, FeatureRepository featureRe
...
-        return modelMapper.map(savedOffer, OfferDto.class);
+        User currentUser = authService.getCurrentUser();
+        Role currentUserRole = currentUser.getRole();
+        if (currentUserRole.hasOfferCreatePrivilege()) {
+            Offer offer = modelMapper.map(offerDto, Offer.class);
+            offer.setOwner(currentUser);
+            Offer savedOffer = offerRepository.save(offer);
+            return modelMapper.map(savedOffer, OfferDto.class);
+        } else {
+            throw new NoAccessToOfferOperationException("Access denied, log as Place Owner or System Admin");
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: You can simplify this line by calling no-argument constructor of the exception:

```
throw new NoAccessToOfferOperationException()
```

and then by encapsulation your string ```"Access denied, log as Place Owner or System Admin"``` in the exception class:

```
public class NoAccessToOfferOperationException extends RuntimeException {

    public NoAccessToOfferOperationException() {
        super("Access denied, log as Place Owner or System Admin");
    }
}
```

By doing this you can re-use ```NoAccessToOfferOperationException``` exception and throw it in multiple places without remembering to provide the same ```String```.

File: src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java [hunk 8]
```diff
@@ -47,23 +50,27 @@ public OfferDto getById(long id) {
...
 
     public OfferDto updateOffer(long id, CreateOfferDto offerDto) {
         Optional<Offer> optionalOffer = offerRepository.findById(id);
-        if (authService.isPoLoggedIn() || authService.isAdminLoggedIn()) {
-            Offer offer = optionalOffer.orElseThrow(
-                    () -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Offer with id=" + id + " not found"));
+        User currentUser = authService.getCurrentUser();
+        Role currentUserRole = currentUser.getRole();
+        Offer offer = optionalOffer.orElseThrow(
+                () -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Offer with id=" + id + " not found"));
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: **warn** I can see that business services are still coupled to implementation details of HTTP protocol in some places. Custom exception should be thrown instead (+ custom exception mapper should be added).

**Review by mgjvm — APPROVED** [review]

File: src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java (hunk 8 above)
Comment (emstryszawski) [reply src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: They will be edited as I did in addOffer method, I wanted to make commit with not finished service just to show my colleagues a way how to map exceptions.

File: src/main/java/dk/bec/polonez/reservationsystem/controller/OfferController.java [hunk 9]
```diff
@@ -42,12 +42,12 @@ public OfferDto getOfferById(@PathVariable long id) {
...
+    public ResponseEntity<OfferDto> deleteOffer(@PathVariable long id) {
+        return new ResponseEntity<>(offerService.deleteOffer(id), HttpStatus.OK);
     }
 
-    @PostMapping("{id}/features")
-    public ResponseOfferFeatureDto addFeatureToOffer(@RequestParam long featureId, @PathVariable long offerId) {
-        return offerService.addFeatureToOffer(featureId, offerId);
+    @PostMapping("{offerId}/features")
+    public ResponseEntity<ResponseOfferFeatureDto> addFeatureToOffer(@RequestParam long featureId, @PathVariable long offerId) {
+        return new ResponseEntity<>(offerService.addFeatureToOffer(featureId, offerId), HttpStatus.OK);
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/controller/OfferController.java]: **info** According to REST conventions POST should return 201 (not 200) after a resource is successfully added to a collection.

**Review by mgjvm — APPROVED** [review]

File: src/main/java/dk/bec/polonez/reservationsystem/controller/OfferController.java [hunk 10]
```diff
@@ -30,22 +32,22 @@ public OfferDto getOfferById(@PathVariable long id) {
...
-    public OfferDto addOffer(@RequestBody CreateOfferDto offerDto) {
-        return offerService.addOffer(offerDto);
+    public ResponseEntity<OfferDto> addOffer(@Valid @RequestBody CreateOfferDto offerDto) {
+        return new ResponseEntity<>(offerService.addOffer(offerDto), HttpStatus.CREATED);
     }
 
     @PutMapping("{id}")
-    public OfferDto updateOffer(@PathVariable long id, @RequestBody CreateOfferDto offerDto) {
-        return offerService.updateOffer(id, offerDto);
+    public ResponseEntity<OfferDto> updateOffer(@PathVariable long id,@Valid @RequestBody CreateOfferDto offerDto) {
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/controller/OfferController.java]: **advice** Bad code formatting (ALT + CTRL + L in IntelliJ).

File: src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java [hunk 11]
```diff
@@ -35,32 +42,76 @@ public OfferService(OfferRepository offerRepository, FeatureRepository featureRe
...
+        Feature feature = optionalFeature.orElseThrow(
+                () -> new NotFoundObjectException(Feature.class, featureId));
+
+        User currentUser = authService.getCurrentUser();
+        Role currentUserRole = currentUser.getRole();
+
+        if (!currentUserRole.hasOfferUpdatePrivilege())
+            throw new NoAccessToOfferOperationException();
+        if (!offer.getOwner().equals(currentUser) && !currentUserRole.hasOfferDeleteOthersOfferPrivilege)
+            throw new NoAccessToOfferOperationException("You cannot update other Place Owner's offer features as a Place Owner, to do that log as an Admin");
```
Comment (mgjvm) [inline src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: **advice** Exception class name should explain what this exception is about business-wise - and then a ```String``` message (with explanation what happened) will become redundant and you will be able to remove it.

Generally speaking, if you're re-using single exception but you throw it in different places in the source code with different ```String``` messages, then you can consider defining seprate exception for each case (= no reuse). Each separate exception would have single responsibility

**Review by mgjvm — APPROVED** [review]

File: src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java [hunk 12]
```diff
@@ -35,32 +42,76 @@ public OfferService(OfferRepository offerRepository, FeatureRepository featureRe
...
-        Feature feature = featureRepository.getById(featureId);
-        Offer offer = offerRepository.getById(offerId);
+        Optional<Feature> optionalFeature = featureRepository.findById(featureId);
+        Optional<Offer> optionalOffer = offerRepository.findById(offerId);
+        Offer offer = optionalOffer.orElseThrow(
+                () -> new NotFoundObjectException(Offer.class, offerId));
+        Feature feature = optionalFeature.orElseThrow(
+                () -> new NotFoundObjectException(Feature.class, featureId));
+
+        User currentUser = authService.getCurrentUser();
```
Comment (MaciejY0R5) [inline src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: logika sprawdzajaca privilege powinna zostac przeniesiona do innej klasy, odpowiedzialnej za sprawdzanie dostepow

File: src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java [hunk 13]
```diff
@@ -35,32 +42,76 @@ public OfferService(OfferRepository offerRepository, FeatureRepository featureRe
...
+        Role currentUserRole = currentUser.getRole();
+
+        if (!currentUserRole.hasOfferUpdatePrivilege())
+            throw new NoAccessToOfferOperationException();
+        if (!offer.getOwner().equals(currentUser) && !currentUserRole.hasOfferDeleteOthersOfferPrivilege)
```
Comment (MaciejY0R5) [inline src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java]: hasOfferDeleteOthersOfferPrivilege <- ciezko zrozumiec co oznacza ta nazwa

**Review by MaciejY0R5 — APPROVED** [review]

**Pull request merged**